[android] Use emulated thread-local storage for API 28 and earlier

finagolfin opened the pull request:
[Android before API 29](https://android.googlesource.com/platform/bionic/+/HEAD/android-changes-for-ndk-developers.md#elf-tls-available-for-api-level-29) and [a few other platforms don't support native TLS](https://github.com/llvm/llvm-project/blob/7672216ed7f480c8d461a2d046a74453307f6298/llvm/include/llvm/TargetParser/Triple.h#L1094), so fall back to LLVM's emulated TLS there, [just like clang does](https://github.com/llvm/llvm-project/blob/93caee17add0c7bc6770365b1d3cae93f258d866/clang/lib/Driver/ToolChains/Clang.cpp#L6745). Also, make sure `-Xcc -f{no-,}emulated-tls` flags passed in are applied to control what the Swift compiler does.

Swift 6 [added a thread-local variable to the stdlib for all linux-based platforms](https://github.com/swiftlang/swift/pull/71383/files#diff-4a4beab0106e7043768f49f30baf61eb45116edabc020f16d39259ec2f9b6c6aR26) for the first time, ie outside [linux threading](https://github.com/swiftlang/swift/blob/ee5efff5b17fddfb0e3535acf71ab3d2b80393e9/include/swift/Threading/Impl/Linux/ulock.h#L49) (Android uses pthreads instead), which then generated code in Foundation with native TLS even for Android API 24, breaking my Android CI that runs cross-compiled tests in the Android emulator at API 24.

I assumed the LLVM codegen backend automatically handled this, but it turns out it has to be explicitly enabled by the clang frontend, so I'm doing the same with the Swift frontend now. I just ran the compiler validation suite and toolchain tests on the Nov. 20 trunk snapshot built natively on Android 13 AArch64: this patch fixed running the Foundation tests and looks like it didn't cause any test regressions.

Considering Swift doesn't appear to have a way to choose native TLS for Swift variables, this pull appears to only really be useful for inline functions from C/C++ headers that are inserted into Swift's IR and codegened together.

@azoy, who added this TLS variable in the C/C++ header earlier this year, and @al45tair, let me know what you think.

@compnerd and @hyp, I don't know if you test your Android SDK on API 28 or earlier, let me know if you've seen this regression.

@3405691582, OpenBSD is listed as one of the platforms using emulated TLS in that second link, let us know if that's correct.

finagolfin commented:
@swift-ci please test

compnerd commented on lib/IRGen/IRGen.cpp:
```diff
@@ -148,6 +148,9 @@ swift::getIRTargetOptions(const IRGenOptions &Opts, ASTContext &Ctx) {
     TargetOpts.ThreadModel = llvm::ThreadModel::Single;
 
+  if (Clang->getTargetInfo().getTriple().hasDefaultEmulatedTLS())
+    TargetOpts.EmulatedTLS = true;
```
We probably should honour `-Xcc -femulated-tls` as well.

finagolfin replied:
I chose the established `-Xcc` flags to control this, because this only affects inline functions from C/C++ headers that have been spliced into Swift IR anyway, as Swift doesn't seem to have a TLS annotation for its own variables.

finagolfin committed:
[android] Use emulated thread-local storage for API 28 and earlier

Android before API 29 and a few other platforms don't support native TLS, so
fall back to LLVM's emulated TLS there, just like clang does. Also, make sure
`-Xcc -f{no-,}emulated-tls` flags passed in are applied to control what the
Swift compiler does.

finagolfin commented:
Rebased and updated to add support for command-line flags that manually control this backend codegen flag. I only ran this new test natively on Android, so let's see how well it works on other CI platforms.

@swift-ci please test

finagolfin commented:
@swift-ci please test

finagolfin commented:
@swift-ci please test

finagolfin commented:
@swift-ci please test windows

al45tair commented on lib/IRGen/IRGen.cpp:
```diff
@@ -148,6 +148,8 @@ swift::getIRTargetOptions(const IRGenOptions &Opts, ASTContext &Ctx) {
   if (Clang->getTargetInfo().getTriple().isOSBinFormatWasm())
     TargetOpts.ThreadModel = llvm::ThreadModel::Single;
 
+  TargetOpts.EmulatedTLS = Opts.UseEmulatedTLS;
```
As @compnerd suggested, maybe we should copy the setting from Clang's options? A simple way might be do to something like this (untested, written in this comment):

```suggestion
  TargetOpts.EmulatedTLS = Opts.UseEmulatedTLS || Clang->getCodeGenOpts().EmulatedTLS;
```

finagolfin replied:
As noted above yesterday, I already implemented support for that flag then. However, I just tried your suggestion and that works too and is simpler, so thanks for the suggestion: I will update this pull to reuse your suggested ClangImporter flag instead.

finagolfin replied:
Done

finagolfin commented:
@swift-ci please test

finagolfin commented:
@swift-ci please test

finagolfin commented:
Ping, passed CI and addressed all reviews, just waiting for final approval.

finagolfin commented:
Ping @compnerd, please review, would like to get this into 6.1 and 6.0 next.

compnerd commented on test/IRGen/emulated-tls.swift:
```diff
@@ -0,0 +1,15 @@
+// EMUTLS-macosx: __emutls_get_address
+// EMUTLS-openbsd: __emutls_get_address
+// EMUTLS-windows-msvc: __emutls_get_address
+// EMUTLS-wasi-NOT: __emutls_get_address
```
Hmm, why is this negated for WASI? Perhaps we should warn about the ignored setting?

finagolfin replied:
I honestly didn't want to deal with WASI at all, but was forced to because the linux CI builds it and runs these tests for it too.

My understanding is that the initial `__emutls_v._swift_stdlib_gettid.tid` symbol is swapped in because the LLVM codegen pass runs on all platforms, but it doesn't have `__emutls_get_address` in compiler-rt, like the other platforms do, so it doesn't call that function. I could've simply disabled this test for WASI instead, but I figured it was worth checking the `__emutls_v._swift_stdlib_gettid.tid` symbol at least on all platforms.

This is just an artifact of the current CI and WASI support, I don't think it's worth addressing in the test.

compnerd replied:
Ah, okay, its just that the call is something else? That seems reasonable enough then. Perhaps an issue to track restoring this would be good.

finagolfin replied:
> Ah, okay, its just that the call is something else?

I don't know: I just did what I had to do to get the test to pass, as the CI was forcing me to address WASI but I'm not that familiar with it, didn't know if it even had threads yet. Looking at the relevant LLVM source now, [the `LowerToTLSEmulatedModel()` method that inserts the `__emutls_get_address` call](https://github.com/llvm/llvm-project/blob/112793a90ec84c31ced9bdacad7ce204e2148d2c/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L10512) is not invoked for the WASM target, as it is [for several other architectures](https://github.com/llvm/llvm-project/blob/1729e6e742ba9f6f210550000ace4bec72530c2e/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L10344). It appears that [WASI threads is still in the standardization phase](https://github.com/WebAssembly/proposals?tab=readme-ov-file#phase-4---standardize-the-feature-wg), so the implementation may still be in flux.

@kateinoigakukun, could you let us know if this might be a useful flag for WASM one day and if this test is okay for now?

kateinoigakukun replied:
All WebAssembly targets currently have proper TLS support (assuming dynamic linking is only for Emscripten, not with WASI), so the lowering does not implement emulation at all and silently ignores the option today as you mentioned. The thread representation model in Wasm is still unstable but I don't expect we need the emulation in the future, so it's fine as is now.

compnerd reviewed (approved):
The warning on the ignored emulated TLS would be nice along with an explanation of why its ignored on WASI.

finagolfin commented:
Merging as I want to get this into the release branches next: @compnerd, if you have any further concerns about the test, let me know and we'll get that in.
